opts/opts_test.go

```diff
@@ -54,3 +47,321 @@ func TestNewOpts(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name       string
+		reader     io.Reader
+		wantLoaded bool
+		errType    error
+	}{
+		{
+			name:       "Successful loading",
+			reader:     bytes.NewBuffer(goodSrc),
+			wantLoaded: true,
+			errType:    nil,
+		},
+		{
+			name:       "No source",
+			reader:     nil,
+			wantLoaded: false,
+			errType:    ErrNonNil,
+		},
+		{
+			name:       "Bad source",
+			reader:     bytes.NewBufferString("bad"),
+			wantLoaded: false,
+			errType:    &json.SyntaxError{},
+		},
+		{
+			name:       "Bad content",
+			reader:     bytes.NewBuffer(badSrc),
+			wantLoaded: false,
+			errType:    Error(""),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			opts := &Opts{}
+
+			loader := WithSecurity(tt.reader)
+			err := loader(opts)
+
+			isLoaded := opts.Security != Security{}
+			if isLoaded != tt.wantLoaded {
+				t.Errorf("Security loaded = %v, want Security to be loaded = %v", isLoaded, tt.wantLoaded)
+			}
+			assert(t, err, tt.errType, nil, nil)
+		})
+	}
+}
+
+func TestWithHostCfg(t *testing.T) {
+	goodSrc, err := os.ReadFile("testdata/host_good_all_set.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	badSrc, err := os.ReadFile("testdata/host_bad_unset.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		name       string
+		reader     io.Reader
+		wantLoaded bool
+		errType    error
+	}{
+		{
+			name:       "Successful loading",
+			reader:     bytes.NewBuffer(goodSrc),
+			wantLoaded: true,
+			errType:    nil,
+		},
+		{
+			name:       "No source",
+			reader:     nil,
+			wantLoaded: false,
+			errType:    ErrNonNil,
+		},
+		{
+			name:       "Bad source",
+			reader:     bytes.NewBufferString("bad"),
+			wantLoaded: false,
+			errType:    &json.SyntaxError{},
+		},
+		{
+			name:       "Bad content",
+			reader:     bytes.NewBuffer(badSrc),
+			wantLoaded: false,
+			errType:    Error(""),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			opts := &Opts{}
+
+			loader := WithHostCfg(tt.reader)
+			err := loader(opts)
+
+			isLoaded := opts.HostCfg != HostCfg{}
+			if isLoaded != tt.wantLoaded {
+				t.Errorf("HostCfg loaded = %v, want HostCfg to be loaded = %v", isLoaded, tt.wantLoaded)
+			}
+			assert(t, err, tt.errType, nil, nil)
+		})
+	}
+}
+
+type badReader struct{}
+
+var errBadRead = errors.New("bad read")
+
+func (r badReader) Read(p []byte) (int, error) {
+	return 0, errBadRead
+}
+
+func TestWithSigningRootCert(t *testing.T) {
+	goodCertPEM, err := os.ReadFile("testdata/cert.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	multipleCertsPEM, err := os.ReadFile("testdata/certs.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	badCertPEM, err := os.ReadFile("testdata/cert_bad_cert.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	keyPEM, err := os.ReadFile("testdata/key.pem")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	badPEM, err := os.ReadFile("testdata/cert_bad_pem.pem")
```

Comment:
Isn't it possible to replace all the calls with `os.Open` which returns an `io.Reader`? Do we strictly need the bytes buffer?

Alternative you could do like this 
```go
func readFile(s string) io.Reader {
	f, err := os.ReadFile(s)
	if err != nil {
		t.Fatal(err)
	}
	return bytes.NewBuffer(f)
}

....

		{
			name:       "Successful loading",
			reader:     readFile(goodCertPEM),
			wantLoaded: true,
			errType:    nil,
		},
```
Or something similar. Prevents the duplicate ReadFile segments and I find it a bit cleaner when extending test cases in the future.

Reply:
Good one :+1: